File: note-python/number-of-cases/permutation.py
def permutation(arr, r):
    arr = sorted(arr)
    used = [False for _ in range(len(arr))]
    result = []

    def generate(chosen, used):
        if len(chosen) == r:
            result.append(tuple(chosen))
            return

        for i in range(len(arr)):
            if not used[i] and (i == 0 or arr[i - 1] != arr[i] or used[i - 1]):
                chosen.append(arr[i])
                used[i] = True
                generate(chosen, used)
                used[i] = False
                chosen.pop()

    generate([], used)
    return result


if __name__ == "__main__":
    arr = ["a", "b", "c", "d"]  # "ABAC"
    r = 2

    print(permutation(arr, r))

File: leetcode/note/permutation.js
function permutation(arr, r) {
  arr.sort();
  let used = Array(arr.length).fill(false);
  let result = [];

  function generate(chosen, used) {
    if (chosen.length === r) {
      result.push([...chosen]);
      return;
    }

    for (let i = 0; i < arr.length; i++) {
      if (!used[i] && (i === 0 || arr[i - 1] !== arr[i] || used[i - 1])) {
        chosen.push(arr[i]);
        used[i] = true;
        generate(chosen, used);
        used[i] = false;
        chosen.pop();
      }
    }
  }

  generate([], used);
  return result;
}
